File: leetcode/1203-sort-items-by-groups-respecting-dependencies/Solution.java
import java.util.*;

class Solution {
    public int[] sortItems(int n, int m, int[] group, List<List<Integer>> beforeItems) {
        int[] g = group.clone();
        int nextGroup = m;
        for (int i = 0; i < n; i++) {
            if (g[i] == -1) {
                g[i] = nextGroup++;
            }
        }
        int totalGroups = nextGroup;

        List<List<Integer>> itemGraph = new ArrayList<>();
        for (int i = 0; i < n; i++) itemGraph.add(new ArrayList<>());
        int[] itemInDegree = new int[n];

        List<List<Integer>> groupGraph = new ArrayList<>();
        for (int i = 0; i < totalGroups; i++) groupGraph.add(new ArrayList<>());
        int[] groupInDegree = new int[totalGroups];
        Set<Long> groupEdgesSeen = new HashSet<>();

        for (int i = 0; i < n; i++) {
            for (int dep : beforeItems.get(i)) {
                itemGraph.get(dep).add(i);
                itemInDegree[i]++;

                int gi = g[i], gd = g[dep];
                if (gi != gd) {
                    long key = (long) gd * 100000 + gi;
                    if (groupEdgesSeen.add(key)) {
                        groupGraph.get(gd).add(gi);
                        groupInDegree[gi]++;
                    }
                }
            }
        }

        List<Integer> groupOrder = topoSort(totalGroups, groupGraph, groupInDegree);
        if (groupOrder == null) return new int[0];

        List<Integer> itemOrder = topoSort(n, itemGraph, itemInDegree);
        if (itemOrder == null) return new int[0];

        Map<Integer, List<Integer>> itemsByGroup = new HashMap<>();
        for (int item : itemOrder) {
            itemsByGroup.computeIfAbsent(g[item], k -> new ArrayList<>()).add(item);
        }

        int[] result = new int[n];
        int idx = 0;
        for (int gid : groupOrder) {
            List<Integer> items = itemsByGroup.getOrDefault(gid, Collections.emptyList());
            for (int item : items) {
                result[idx++] = item;
            }
        }

        return result;
    }

    private List<Integer> topoSort(int size, List<List<Integer>> graph, int[] inDegree) {
        Deque<Integer> queue = new ArrayDeque<>();
        for (int v = 0; v < size; v++) {
            if (inDegree[v] == 0) queue.add(v);
        }
        List<Integer> order = new ArrayList<>();
        while (!queue.isEmpty()) {
            int v = queue.poll();
            order.add(v);
            for (int nxt : graph.get(v)) {
                if (--inDegree[nxt] == 0) queue.add(nxt);
            }
        }
        return order.size() == size ? order : null;
    }
}

File: leetcode/1203-sort-items-by-groups-respecting-dependencies/solution.cpp
#include <vector>
#include <queue>
#include <unordered_map>
#include <set>
using namespace std;

class Solution {
public:
    vector<int> sortItems(int n, int m, vector<int>& group, vector<vector<int>>& beforeItems) {
        vector<int> g = group;
        int nextGroup = m;
        for (int i = 0; i < n; i++) {
            if (g[i] == -1) g[i] = nextGroup++;
        }
        int totalGroups = nextGroup;

        vector<vector<int>> itemGraph(n);
        vector<int> itemInDegree(n, 0);
        vector<vector<int>> groupGraph(totalGroups);
        vector<int> groupInDegree(totalGroups, 0);
        set<pair<int,int>> groupEdgesSeen;

        for (int i = 0; i < n; i++) {
            for (int dep : beforeItems[i]) {
                itemGraph[dep].push_back(i);
                itemInDegree[i]++;

                int gi = g[i], gd = g[dep];
                if (gi != gd) {
                    auto key = make_pair(gd, gi);
                    if (!groupEdgesSeen.count(key)) {
                        groupEdgesSeen.insert(key);
                        groupGraph[gd].push_back(gi);
                        groupInDegree[gi]++;
                    }
                }
            }
        }

        auto topoSort = [](int size, vector<vector<int>>& graph, vector<int>& inDegree) -> vector<int> {
            queue<int> q;
            for (int v = 0; v < size; v++) {
                if (inDegree[v] == 0) q.push(v);
            }
            vector<int> order;
            while (!q.empty()) {
                int v = q.front(); q.pop();
                order.push_back(v);
                for (int nxt : graph[v]) {
                    if (--inDegree[nxt] == 0) q.push(nxt);
                }
            }
            if ((int)order.size() != size) return {};
            return order;
        };

        vector<int> groupOrder = topoSort(totalGroups, groupGraph, groupInDegree);
        if ((int)groupOrder.size() != totalGroups) return {};

        vector<int> itemOrder = topoSort(n, itemGraph, itemInDegree);
        if ((int)itemOrder.size() != n) return {};

        unordered_map<int, vector<int>> itemsByGroup;
        for (int item : itemOrder) {
            itemsByGroup[g[item]].push_back(item);
        }

        vector<int> result;
        result.reserve(n);
        for (int gid : groupOrder) {
            if (itemsByGroup.count(gid)) {
                for (int item : itemsByGroup[gid]) {
                    result.push_back(item);
                }
            }
        }

        return result;
    }
};
